lib/ClangImporter/SwiftDeclSynthesizer.cpp

```diff
@@ -2529,3 +2529,131 @@ SwiftDeclSynthesizer::makeDefaultArgument(const clang::ParmVarDecl *param,
 
   return callExpr;
 }
+
+// MARK: C++ foreign reference type constructors
+
+clang::CXXMethodDecl *
+SwiftDeclSynthesizer::synthesizeStaticFactoryForCXXForeignRef(
+    clang::CXXRecordDecl *cxxRecordDecl, clang::ASTContext &clangCtx,
+    clang::Sema &clangSema) {
+
+  clang::QualType cxxRecordTy = clangCtx.getRecordType(cxxRecordDecl);
+
+  clang::FunctionDecl *operatorNew = nullptr;
+  clang::FunctionDecl *operatorDelete = nullptr;
+  bool passAlignment = false;
+  clangSema.FindAllocationFunctions(
+      cxxRecordDecl->getLocation(), clang::SourceRange(), clang::Sema::AFS_Both,
+      clang::Sema::AFS_Both, cxxRecordTy,
+      /*IsArray*/ false, passAlignment, clang::MultiExprArg(), operatorNew,
+      operatorDelete);
+  if (!operatorNew || operatorNew->isDeleted() ||
+      operatorNew->getAccess() == clang::AS_private ||
+      operatorNew->getAccess() == clang::AS_protected)
+    return nullptr;
+
+  clang::CXXConstructorDecl *defaultCtorDecl = nullptr;
+  for (clang::CXXConstructorDecl *ctor : cxxRecordDecl->ctors()) {
+    if (ctor->isDefaultConstructor() && !ctor->isDeleted() &&
+        ctor->getAccess() != clang::AS_private &&
+        ctor->getAccess() != clang::AS_protected) {
+      bool ctorHasDefaultArgs = false;
+      for (const clang::ParmVarDecl *ctorParam : ctor->parameters()) {
+        if (ctorParam->hasDefaultArg()) {
+          ctorHasDefaultArgs = true;
+        }
+      }
+      if (!ctorHasDefaultArgs) {
+        defaultCtorDecl = ctor;
+        break;
+      }
+    }
+  }
+  if (!defaultCtorDecl)
+    return nullptr;
+
+  clang::QualType cxxRecordPtrTy = clangCtx.getPointerType(cxxRecordTy);
+  // Adding `_Nonnull` to the return type of synthesized static factory
+  clangSema.CheckImplicitNullabilityTypeSpecifier(
+      cxxRecordPtrTy, clang::NullabilityKind::NonNull,
+      cxxRecordDecl->getLocation(),
+      /*isParam=*/false,
+      /*OverrideExisting=*/true);
+
+  clang::IdentifierTable &clangIdents = clangCtx.Idents;
+  clang::IdentifierInfo *funcNameToSynthesize = &clangIdents.get(
+      ("__returns_" + cxxRecordDecl->getNameAsString()).c_str());
+  clang::FunctionProtoType::ExtProtoInfo EPI;
+  clang::QualType funcTypeToSynthesize =
+      clangCtx.getFunctionType(cxxRecordPtrTy, {}, EPI);
+
+  clang::CXXMethodDecl *synthesizedCxxMethodDecl = clang::CXXMethodDecl::Create(
+      clangCtx, cxxRecordDecl, cxxRecordDecl->getLocation(),
+      clang::DeclarationNameInfo(funcNameToSynthesize,
+                                 cxxRecordDecl->getLocation()),
+      funcTypeToSynthesize, clangCtx.getTrivialTypeSourceInfo(cxxRecordTy),
```

Comment:
You should create a `TypeSourceInfo` from the function type, not the record type.